Documentation Updates for v2 (#125)
redhat-cop/rhel-edge-automation-arch: docs/basic-walkthrough.md

# Basic Walkthrough

This guide will help familiarize yourself with the process of building your first RHEL for Edge image using this architecture. By the end of this walk-through, you will:

* Understand the primary components in the architecture
* Build a RHEL for Edge image from a Blueprint
* Publish a Kickstart file referencing the previously built RHEL for Edge image

## Prerequisites

The following requirements must be satisfied prior to beginning the walk-through:

1. OpenShift CLI Tool
2. Tekton CLI Tool
3. An OpenShift cluster provisioned with the tooling associated in this repository
3. Access to the OpenShift cluster as a user with `cluster-admin` access.

## Use Case Overview

This walk-through will illustrate the ease of building, publishing and consuming RHEL for Edge content. For the sample use case, an edge node with the [IBM Developer Model Asset Exchange: Weather Forecaster](https://github.com/IBM/MAX-Weather-Forecaster) application running in a container will be built and deployed. This process consists of the following:

> * Run a series of pipelines that: 

Reviewer: ```suggestion
* Execute a series of pipelines that: 
```

  + Use Image Builder to create a custom RHEL for Edge image (OSTree commit) using compose image type `rhel-edge-container`

>   + Push generated OCI tar archive to Quay

Reviewer: tar archive?

Author: I was thinking literally the tarball pulled from `composer-cli compose image`, I'll just change it to 'OCI container archive'.

  + Deploy OCI container on OpenShift for staging

>   + Synchronize OStree content in web server running on OpenShift for production

Reviewer: ```suggestion
  + Synchronize OStree content from web server running on OpenShift for production promotion
```

* `-s rfe-automation` - The name of the Service Account used to run the pipeline.
* `--use-param-default` - The default Pipeline parameters will be applied unless explicitly specified.
* `-p blueprint-dir=hello-world` - The directory containing the blueprint file in the cloned repository. By default, the _blueprints_ branch of this repository will be used.

The output of the command will provide a command to view the progress of the build.

_Note: The process of building a RHEL for Edge image takes time!_

### Verification

Once the pipeline completes, the OCI container generated by Image Builder should be stored in Quay. To verify, obtain the route to Quay by running the following command:

```shell
oc get quayregistry quay -n quay -ojsonpath='{.status.registryEndpoint}'
```

> Quay is not setup to use OpenShift's SSO, so the username can be found by running:

Reviewer: ```suggestion
Quay is not setup to use external authentication, so the username can be found by running:
```

Quay can only handle OIDC for which OCP only exposes an OAuth server

```shell
oc get secret quay-rfe-setup -n rfe -o go-template='{{ .data.username | base64decode }}'
```

And the password by running:

```shell
oc get secret quay-rfe-setup -n rfe -o go-template='{{ .data.password | base64decode }}'
```

> Once logged in, click on the RFE organization to the right of the page under _Users and Organizations_ and then select the repository name associated with your blueprint.

Reviewer: ```suggestion
Once logged in to Quay, click on the RFE organization to the right of the page under _Users and Organizations_ and then select the repository name associated with your blueprint.
```

To the left of the screen, click the _Tags_ icon to view associated tags. Each pipeline run will create two tags:

* A _latest_ tag that points to the most recent image.
* A tag with the version specified in the blueprint.

At this point, you could manually pull/deploy the container for use in the deployment of RFE content.

### Pipeline Results

Each pipeline run returns three results:

* `build-commit` - The Build Commit ID from Image Builder
* `image-path` - Location in Quay registry of OCI container
* `image-tags` - Tags applied to the container (JSON list)

To view the results, find the latest pipeline run. Use the following command as an example:

```shell
$ tkn pipelinerun list -n rfe --label tekton.dev/pipeline=rfe-oci-image-pipeline --limit 1
NAME                               STARTED     DURATION     STATUS
rfe-oci-image-pipeline-run-2lpwc   1 day ago   13 minutes   Succeeded
```

Then run the following to view the pipeline results:

```shell

> $ oc get pipelinerun rfe-oci-image-pipeline-run-2lpwc -ojsonpath='{.status.pipelineResults}' | jq .

Reviewer: Do we want to introduce `jq` as a dependency for running through the tutrial

[
  {
    "name": "build-commit",
    "value": "ab07f144-43a7-49b3-93de-99e1562435f9"
  },
  {
    "name": "image-path",
    "value": "quay-quay-quay.apps.cluster.com/rfe/hello-world"
  },
  {
    "name": "image-tags",
    "value": "[\"latest\", \"0.0.1\"]"
  }
]
```

## Staging OCI Container with OSTree Commit

> Now that we have our OCI container from Image Builder with our OSTree Commit, we can run a pipeline to deploy it as a staging environment in OpenShift.

Reviewer: ```suggestion
Now that we have our OCI container from Image Builder with our OSTree Commit in Quay, we can run a pipeline to deploy it as a staging environment in OpenShift.
```

From the root of the project, execute the following command to instantiate the `rfe-oci-stage-pipeline` to deploy the OCI container built in the previous pipeline (`rfe-oci-image-pipeline`) run.

```shell
tkn pipeline start rfe-oci-stage-pipeline \
--workspace name=shared-workspace,volumeClaimTemplateFile=openshift/resources/pipelines/volumeclaimtemplate.yaml \
-s rfe-automation \
--use-param-defaults \
-p image-path=$(oc get quayregistry quay -n quay -ojsonpath='{.status.registryEndpoint}')/rfe/hello-world \
-p image-tag=latest
```

This command is similar to the previous pipeline run, but the following parameters are used:

> * `-p image-path=...` - The path to the OCI container stored in the Quay registry.

Reviewer: Do we want to give an example here instead of `...`? Like a fake example and elaborate that users should replace with their own

Each pipeline run returns one result:

* `content-path` - The path to the OSTree repository.

To view the results, find the latest pipeline run. Use the following command as an example:

```shell
$ tkn pipelinerun list --label tekton.dev/pipeline=rfe-oci-stage-pipeline --limit 1
NAME                               STARTED     DURATION     STATUS
rfe-oci-stage-pipeline-run-cxkxq   1 day ago   13 minutes   Succeeded
```

Then run the following to view the pipeline results:

```shell
$ oc get pipelinerun -n rfe rfe-oci-stage-pipeline-run-cxkxq -ojsonpath='{.status.pipelineResults}' | jq .
[
  {
    "name": "content-path",
    "value": "http://hello-world-latest-rfe.apps.cluster.com/repo"
  }
]
```

## Moving from Stage to Production

The next stage involves synchronizing our OSTree Commit from our staging environment to production.

From the root of the project, execute the following command to instantiate the `rfe-oci-publish-content-pipeline` pipeline:

```shell
tkn pipeline start rfe-oci-publish-content-pipeline \
--workspace name=shared-workspace,volumeClaimTemplateFile=openshift/resources/pipelines/volumeclaimtemplate.yaml \
-s rfe-automation \
--use-param-defaults \

> -p image-path=(oc get route -n quay quay-quay -ojson | jq -r .status.ingress[].host)/rfe/hello-world \

Reviewer: cant we use `.spec.host` instead of the dependency on jq?

-p image-tag=latest 
```

This command is similar to the previous pipeline run, but the following parameters are used:

* `-p image-path=...` - The path to the OCI container stored in the Quay registry.
* `-p image-tag=latest` - Use the image with the tag _latest_.

### Verification

> Once the pipeline runs, the OSTree Commit is synchronized to the production webserver. Verify the hash by running the following command:

Reviewer: Elaborate on what you mean by `hash`.
